### PR TITLE
Update ROS params from another node

### PR DESCRIPTION
**Pull request by steph-lheureux**
# Summary
This PR requests the merge of code which provides functionality for nodes to modify each other's [ROS parameters](https://docs.ros.org/en/humble/Concepts/Basic/About-Parameters.html) using [services](https://docs.ros.org/en/humble/Tutorials/Beginner-CLI-Tools/Understanding-ROS2-Services/Understanding-ROS2-Services.html). The `RemoteParamsClient` class streamlines setting another node's parameters to a user-friendly abstraction.
# Changes
### Reasoning
The `RemoteParamsClient` class was created with the primary objective of providing a streamlined process for setting parameters of another node. The class encapsulates underlying complexities of remotely setting parameters, which will become beneficial when we eventually set numerous node's parameters from various nodes across the deck. The abstraction simplifies the process and reduces redundancy which will result in cleaner code and increased ease-of-use.
#### Client node
*The client node is the node which makes a request to update another node's parameters*
1. Import the `RemoteParamsClient` class to any node on the deck
	```py
	from .set_remote_params import RemoteParamsClient
	```
2. Create an instance of the class in the `__init__()` function, and pass it `self` (the current node) and the name of the node of whose parameters to set
	```py
	self.__set_params = RemoteParamsClient(self, "other_node_name")
	```
3. In a callback, choose the parameters to set and their new values by calling `update_params()` with arguments of the name of the parameter to set and its new value. This may be called multiple times to set multiple parameters
	```py
	self.__set_params.update_params("param_name", param_val)
	```
4. Once all the parameter values have been instantiated using `update_params()`, send the parameters via service to the other node by calling `send_params()`
	```py
	self.__set_params.send_params()
	```

## Updated naming of nodes
- `input_xbox_one` handles more than just the xbox input now so it has be renamed to `pilot_input`
- `input_keyboard` is unnecessarily wordy and can be shortened to just `keyboard` without losing meaning

### CODE REVIEW

File: src/seahawk/seahawk_deck/set_remote_params.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        return future.result()                      # Return if the parameter is set correctly
+    
+
+class SrvRemoteParams():
+    """
+    Service node. Abstracts requesting setting node's parameters from another node via service to a class.
+    """
+    def __init__(self, this_node: Node):
+        """
+        Set up request to setting thi node's parameters from another node
```
Comment (steph-lheureux) [inline src/seahawk/seahawk_deck/set_remote_params.py]: Change comment to the following
> Set up request to set this node's parameters from another node

File: src/seahawk/seahawk_deck/keyboard.py [hunk 2]
```diff
@@ -35,60 +35,82 @@
...
+        # Set up client to remotely set parameters on 'pilot_input' node using a service
+        self.__set_params_input_xbox_one = CliRemoteParams(self, "pilot_input")
+
         # Get current user terminal settings and save them for later
         self.__settings = settings
+
+        # Store the current key
+        self.__cur_key = ""
     
     def __get_key(self) -> str:
```
Comment (OrionOth) [inline src/seahawk/seahawk_deck/keyboard.py]: It doesn't look like this function returns anything anymore, consider changing the signature notation to reflect that?

File: src/seahawk/seahawk_deck/keyboard.py [hunk 3]
```diff
@@ -35,60 +35,82 @@
...
             raise KeyboardInterrupt("Terminated the process with ctrl-c")
         
+        self.__pub_callback()
+        self.__param_callback()
+
+    def __pub_callback(self):
+        """
+        Publishes the current key to 'keystroke' topic
+
+        Throws 'KeyboardInterrupt' if user enters ctrl-c
```
Comment (OrionOth) [inline src/seahawk/seahawk_deck/keyboard.py]: I don't believe this comment is accurate anymore; the `KeyboardInterrupt` is thrown above in the `callback` function.

File: src/seahawk/seahawk_deck/keyboard.py [hunk 4]
```diff
@@ -35,60 +35,82 @@
...
         """
-        super().__init__("input_keyboard")
+        super().__init__("keyboard")
     
-        # Create publisher to topic 'key_press'
+        # Create publisher to topic 'keystroke'
         self.__key_pub = self.create_publisher(String, "keystroke", 10)
 
+        # Set up client to remotely set parameters on 'pilot_input' node using a service
+        self.__set_params_input_xbox_one = CliRemoteParams(self, "pilot_input")
```
Comment (OrionOth) [inline src/seahawk/seahawk_deck/keyboard.py]: Given renaming elsewhere, I think you should rename `__set_params_input_xbox_one` and related to something more like `__set_params_pilot_input` for better consistency.

File: src/seahawk/seahawk_deck/keyboard.py [hunk 5]
```diff
@@ -35,60 +35,82 @@
...
+        msg.data = self.__cur_key
         self.__key_pub.publish(msg)
+    
+    def __param_callback(self):
+        """
+        Updates parameters for 'throttle_curve_choice'
+        """
+        # If the user pressed '0', '1', or '2' send that information to the __set_params_input_xbox_one
+        if self.__cur_key in ["0", "1", "2"]:
+            self.__set_params_input_xbox_one.update_params("throttle_curve_choice", self.__cur_key)
```
Comment (OrionOth) [inline src/seahawk/seahawk_deck/keyboard.py]: This is a very unergonomic method of keybinding. If we wanted to change these to, say, letters instead, then the letter would be sent. It's better to abstract this I'd say. Consider using a dict, or the index in the array, and sending numbers 0-3 instead of characters "0"-"3", or using id-like names (`"linear"`, `"gentle"`, `"deadzone-50"`...).

File: src/seahawk/seahawk_deck/pilot_input.py [hunk 6]
```diff
@@ -77,21 +79,30 @@ def reset(self):
...
 
+        # Create publishers and subscriptions
         self.subscription = self.create_subscription(Joy, "joy", self.__callback, 10)
         self.__twist_pub = self.create_publisher(Twist, "desired_twist", 10)
         self.__claw_pub = self.create_publisher(Bool, "claw_state", 10)
-        
+
+        # Create and store parameter which determines which throttle curve
+        # the pilot wants to use named 'throttle_curve_choice'. Defaults to '0'
+        self.declare_parameter("throttle_curve_choice", "0")
```
Comment (OrionOth) [inline src/seahawk/seahawk_deck/pilot_input.py:97]: See my other comment on `keyboard.py` about unergonomic keybinding

**Review by OrionOth — CHANGES_REQUESTED** [review]
I like the node name changes.

I am very much not a fan of the confusing names `SrvRemoteParams` and `CliRemoteParams`. Maybe `RemoteParamsService` and `RemoteParamsClient`? Also, perhaps consider changing some of the wording/comments around the "service" to specify that it isn't so much a service node in itself but rather is a utility class to add a service to a node. `RemoteParamsServiceWrapper`? I do like the way you are implementing it, though.

TLDR; my gripes are around naming conventions mostly

File: src/seahawk/seahawk_deck/keyboard.py (hunk 2 above)
Comment (steph-lheureux) [reply src/seahawk/seahawk_deck/keyboard.py]: Thanks, it returned something at some point and I forgot to remove the type hint. Updated to
```py
def __get_key(self):
```

File: src/seahawk/seahawk_deck/set_remote_params.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+        Returns:
+            SetParametersResult() which lets ROS2 know if the parameters were set correctly or not
+        """
+        try:
+            # Try to set parameters to updated values
+            self.__this_node.set_parameters(params)
+        except:
+            return SetParametersResult(successful=False)
+        return SetParametersResult(successful=True)
```
Comment (OrionOth) [inline src/seahawk/seahawk_deck/set_remote_params.py]: This is effectively re-implementing what ROS is already doing, setting the parameters. This leads to unpredictable behavior, likely recursive.

https://docs.ros.org/en/rolling/Concepts/Basic/About-Parameters.html#parameter-callbacks
> A ROS 2 node can register three different types of callbacks to be informed when changes are happening to parameters. All three of the callbacks are optional.
> ...
> The second is known as a “set parameter” callback, and can be set by calling `add_on_set_parameters_callback` from the node API. The callback is passed a list of immutable `Parameter` objects, and returns an `rcl_interfaces/msg/SetParametersResult`. The main purpose of this callback is to give the user the ability to inspect the upcoming change to the parameter and explicitly reject the change.
> ...

here's an example of what `add_on_set_parameters_callback` is supposed to be doing: https://github.com/ros2/demos/blob/rolling/demo_nodes_cpp/src/parameters/set_parameters_callback.cpp#L70

**Review by OrionOth — CHANGES_REQUESTED** [review]
`SrvRemoteParams` is completely unnecessary here.

File: src/seahawk/seahawk_deck/keyboard.py (hunk 4 above)
Comment (steph-lheureux) [reply src/seahawk/seahawk_deck/keyboard.py]: Fixed
```py
# Send all params to pilot_input node
        self.__set_params_pilot_input.send_params()
```

Comment (steph-lheureux) [reply src/seahawk/seahawk_deck/keyboard.py]: Screenshot from 2024-01-15 18-23-13

File: src/seahawk/seahawk_deck/set_remote_params.py (hunk 7 above)
Comment (steph-lheureux) [reply src/seahawk/seahawk_deck/set_remote_params.py]: Nice catch, I removed the entire `SrvRemoteParams` class and it still works

**Review by OrionOth — APPROVED** [review]
LGTM! Don't forget to make another PR with the commit from `git revert 757e2af` in order to get the keyboard changes too.

File: src/seahawk/seahawk_deck/set_remote_params.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    def __init__(self, this_node: Node, other_node_name: str):
+        """
+        Set up request to set another node's parameters via service named '/other_node_name/set_parameters'
+    
+        Args:
+            this_node: Node which makes the request to set another node's parameters. Client node
+            other_node_name: Name of the node whose parameters should be set. Service node
+        """
+        # Create service name. Services to set parameters must be named as '/other_node_name/set_parameters'
+        srv_name = "/" + other_node_name + "/set_parameters"
```
Comment (OrionOth) [inline src/seahawk/seahawk_deck/set_remote_params.py:51]: I think it would be somewhat "more correct" to not have the prefixing `/` as there is a sort of namespace hierarchy in ROS.

Comment (steph-lheureux) [reply src/seahawk/seahawk_deck/set_remote_params.py:51]: Services are always in the form `/node_name/srv_name`
services

Comment (OrionOth) [reply src/seahawk/seahawk_deck/set_remote_params.py:51]: Yes, but think of it like absolute paths versus relative paths. ROS will prefix it with whatever it needs.
At least, that's my interpretation of how it works.

File: src/seahawk/seahawk_deck/set_remote_params.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        """
+        Appends a single new parameter given its name and new value to the list of parameters to set
+
+        Args: 
+            param_name: Name of the parameter
+            param_value: Value of the parameter (note must of a be a valid parameter type)
+        """
+        self.__param_list.append(Parameter(name=param_name, value=param_value).to_parameter_msg())
+	
+    def send_params(self) -> (Any | None):
```
Comment (mike-matera) [inline src/seahawk/seahawk_deck/set_remote_params.py]: This is great code Steph. It makes it way easier to grab and hold onto a remote parameter that you're likey to change often. Here's something to consider. It looks like users of this have to:

```python 
p.update_params('foo', 'bar')
p.update_params('thing', 'bak')
p.send_params()
```

If you liked, you could combine the calls by making `send_params` take a dictionary:

```python
p.send_params({
  'foo': 'bar', 
  'thing': 'bak',
})
```

Just something to consider.

Comment (OrionOth) [reply src/seahawk/seahawk_deck/set_remote_params.py:75]: This history is going to be rewritten as we discussed a few minutes ago, so this review will be lost. Sorry! Normally I wouldn't ever rewrite history on an active PR but this is emergency mode I think.

**Comment by OrionOth** [comment]
This force push was because all of those commits were undesired by Steph.

**Review by mike-matera — APPROVED** [review]
Looks good!

File: src/seahawk/seahawk_deck/set_remote_params.py (hunk 9 above)
Comment (steph-lheureux) [reply src/seahawk/seahawk_deck/set_remote_params.py]: Thanks!! In the current use cases I can think of, we are only going to be updating one parameter at a time, what would be an instance where we would want to update multiple parameters? I could not think of a case when writing the code, but knew maybe it could be a possibility and thus allowed the user to call `update_params()` multiple times. In my opinion it is not necessary to optimize for an unlikely case but let me know if there is something I am not considering

Comment (OrionOth) [reply src/seahawk/seahawk_deck/set_remote_params.py]: The main issue with parameters is there are only a few fixed types. If you wanted to use the thrust matrix as a parameter, for example, it would have to be treated as an array with separate parameter(s) to describe how many columns/rows it has.

File: src/seahawk/seahawk_deck/pilot_input.py [hunk 10]
```diff
@@ -77,21 +76,27 @@ def reset(self):
...
```
Comment (liamgilligan) [inline src/seahawk/seahawk_deck/pilot_input.py:97]: Suggestion (strong): The parameter is created but not stored. You should create a member variable (such as `self.throttle_curve_choice`) to hold the value of the parameter so that you don't have to call `self.get_parameter("throttle_curve_choice").value` to get the value every time.

File: src/seahawk/seahawk_deck/set_remote_params.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+# Type hints
+from typing import Any
+
+# Parameters
+from rcl_interfaces.srv import SetParameters
+from rclpy.parameter import Parameter
+
+class RemoteParamsClient():
+    """
+    Abstracts requesting to set another node's parameters via service to a class.
```
Comment (liamgilligan) [inline src/seahawk/seahawk_deck/set_remote_params.py]: Nit: change "via service" to "via a service call".

Comment (liamgilligan) [reply src/seahawk/seahawk_deck/set_remote_params.py]: There are multiple instances of this, this nitpick applies to all.

File: src/seahawk/seahawk_deck/set_remote_params.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    """
+    Abstracts requesting to set another node's parameters via service to a class. 
+    """
+    def __init__(self, this_node: Node, other_node_name: str):
+        """
+        Set up request to set another node's parameters via service named '/other_node_name/set_parameters'
+    
+        Args:
+            this_node: Node which makes the request to set another node's parameters. Client node
+            other_node_name: Name of the node whose parameters should be set. Service node
```
Comment (liamgilligan) [inline src/seahawk/seahawk_deck/set_remote_params.py]: Nit: Change "Service node" to "Parameter node", as this class only changes parameters and does not handle service calls broadly.

File: src/seahawk/seahawk_deck/set_remote_params.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+            param_value: Value of the parameter (note must of a be a valid parameter type)
+        """
+        self.__param_list.append(Parameter(name=param_name, value=param_value).to_parameter_msg())
+	
+    def send_params(self) -> (Any | None):
+        """
+        Sends updated parameter list to the service
+
+        Returns:
+            The result set by the task, None if no result was set
```
Comment (liamgilligan) [inline src/seahawk/seahawk_deck/set_remote_params.py]: Nit: Confusing `Returns:` description in the docstring, consider elaborating.

**Review by liamgilligan — APPROVED** [review]
Overall, looks super good. I could only find one real area to change, which is between an Issue and a Suggestion (I marked it as a strong suggestion). Otherwise, I could only find nitpicks. A lot of the stylistic changes you made are super good to see, and the added functionality is awesome.

I'm marking as approved but I would *strongly* prefer you to consider some of the changes I recommended, however code is good as is IMO. Specifically, consider creating a member variable for storing the value of the `throttle_curve_choice` parameter in the `pilot_input` node. 

Great work!

File: src/seahawk/seahawk_deck/set_remote_params.py (hunk 9 above)
Comment (liamgilligan) [reply src/seahawk/seahawk_deck/set_remote_params.py]: Can't help but give a list comprehension for dealing with that:

`self.__req.parameters = [Parameter(name=param_name, value=param_value).to_parameter_msg() for param_name, param_value in self.__param_list.items()]`

Comment (liamgilligan) [reply src/seahawk/seahawk_deck/set_remote_params.py]: That isn't an issue @OrionOth, at least in regards to the motor configuration matrix. The parameter that handles variable center of mass is a three element array that indicates the offset of the center of mass in (x, y, z) (see 694f221).

File: src/seahawk/seahawk_deck/pilot_input.py (hunk 10 above)
Comment (steph-lheureux) [reply src/seahawk/seahawk_deck/pilot_input.py:97]:
```py
# Create and store parameter which determines which throttle curve
# the pilot wants to use named 'throttle_curve_choice'. Defaults to '0'
self.declare_parameter("throttle_curve_choice", "0")
self.__throttle_curve_choice = self.get_parameter("throttle_curve_choice").value
```
Added member variable

File: src/seahawk/seahawk_deck/keyboard.py (hunk 5 above)
Comment (steph-lheureux) [reply src/seahawk/seahawk_deck/keyboard.py]: [wont-fix] This will be updated in the gui implementation of the keyboard, no use fixing something which will be re-written #186

File: src/seahawk/seahawk_deck/pilot_input.py (hunk 6 above)
Comment (steph-lheureux) [reply src/seahawk/seahawk_deck/pilot_input.py:97]: See other comment above

File: src/seahawk/seahawk_deck/pilot_input.py (hunk 10 above)
Comment (OrionOth) [reply src/seahawk/seahawk_deck/pilot_input.py:97]: This isn't set anywhere else.
